core/vm/eof.go

```diff
@@ -0,0 +1,315 @@
+	ErrInvalidVersion         = errors.New("invalid version")
+	ErrMissingTypeHeader      = errors.New("missing type header")
+	ErrInvalidTypeSize        = errors.New("invalid type section size")
+	ErrMissingCodeHeader      = errors.New("missing code header")
+	ErrInvalidCodeHeader      = errors.New("invalid code header")
+	ErrInvalidCodeSize        = errors.New("invalid code size")
+	ErrMissingDataHeader      = errors.New("missing data header")
+	ErrMissingTerminator      = errors.New("missing header terminator")
+	ErrTooManyInputs          = errors.New("invalid type content, too many inputs")
+	ErrTooManyOutputs         = errors.New("invalid type content, too many inputs")
+	ErrInvalidSection0Type    = errors.New("invalid section 0 type, input and output should be zero")
+	ErrTooLargeMaxStackHeight = errors.New("invalid type content, max stack height exceeds limit")
+	ErrInvalidContainerSize   = errors.New("invalid container size")
+)
+
+var eofMagic = []byte{0xef, 0x00}
+
+// HasEOFByte returns true if code starts with 0xEF byte
+func HasEOFByte(code []byte) bool {
+	return len(code) != 0 && code[0] == eofFormatByte
+}
+
+// hasEOFMagic returns true if code starts with magic defined by EIP-3540
+func hasEOFMagic(code []byte) bool {
+	return len(eofMagic) <= len(code) && bytes.Equal(eofMagic, code[0:len(eofMagic)])
+}
+
+// isEOFVersion1 returns true if the code's version byte equals eof1Version. It
+// does not verify the EOF magic is valid.
+func isEOFVersion1(code []byte) bool {
+	return 2 < len(code) && code[2] == byte(eof1Version)
+}
+
+// Container is an EOF container object.
+type Container struct {
+	Types []*FunctionMetadata
+	Code  [][]byte
+	Data  []byte
+}
+
+// FunctionMetadata is an EOF function signature.
+type FunctionMetadata struct {
+	Input          uint8
+	Output         uint8
+	MaxStackHeight uint16
+}
+
+// MarshalBinary encodes an EOF container into binary format.
+func (c *Container) MarshalBinary() []byte {
+	// Build EOF prefix.
+	b := make([]byte, 2)
+	copy(b, eofMagic)
+	b = append(b, eof1Version)
+
+	// Write section headers.
+	b = append(b, kindTypes)
+	b = binary.BigEndian.AppendUint16(b, uint16(len(c.Types)*4))
+	b = append(b, kindCode)
+	b = binary.BigEndian.AppendUint16(b, uint16(len(c.Code)))
+	for _, code := range c.Code {
+		b = binary.BigEndian.AppendUint16(b, uint16(len(code)))
+	}
+	b = append(b, kindData)
+	b = binary.BigEndian.AppendUint16(b, uint16(len(c.Data)))
+	b = append(b, 0) // terminator
+
+	// Write section contents.
+	for _, ty := range c.Types {
+		b = append(b, []byte{ty.Input, ty.Output, byte(ty.MaxStackHeight >> 8), byte(ty.MaxStackHeight & 0x00ff)}...)
+	}
+	for _, code := range c.Code {
+		b = append(b, code...)
+	}
+	b = append(b, c.Data...)
+
+	return b
+}
+
+// UnmarshalBinary decodes an EOF container.
+func (c *Container) UnmarshalBinary(b []byte) error {
+	if !hasEOFMagic(b) {
+		return fmt.Errorf("%w: want %s", ErrInvalidMagic, eofMagic)
```

Comment:
```suggestion
		return fmt.Errorf("%w: want %x", ErrInvalidMagic, eofMagic)
```
Otherwise the error will contain non-printable characters, 
```
err: invalid magic: want <EF>^@
```

Reply:
fixed